File: typescript/src/validate.ts
import { basename } from "node:path";

import Ajv2020, { type ErrorObject } from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

import profileSchema from "../../schema/v1/agent-profile.schema.json";
import deltaSchema from "../../schema/v1/agent-state-delta.schema.json";

import { profileDigests } from "./canonical.js";
import type { AgentProfile, Issue, JsonObject, JsonValue, ValidationReport } from "./types.js";

const ajv = new Ajv2020({ allErrors: true, strict: false });
addFormats(ajv);
const validateProfileSchema = ajv.compile(profileSchema);
const validateDeltaSchema = ajv.compile(deltaSchema);

const ENV_REF = /^\$\{[A-Z][A-Z0-9_]{0,63}\}$/;
const HEADER_REF = /^(Bearer )?\$\{[A-Z][A-Z0-9_]{0,63}\}$/;
const VAR_REF = /\$\{\{\s*vars\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}/g;
const PROFILE_SUFFIXES = [".agent.yaml", ".agent.yml", ".agent.json", ".agent.md"];

const SECRET_PATTERNS: Array<[RegExp, string]> = [
  [/\bsk-[A-Za-z0-9_-]{16,}/, "OpenAI-style API key"],
  [/\bsk-ant-[A-Za-z0-9_-]{16,}/, "Anthropic API key"],
  [/\bgh[pousr]_[A-Za-z0-9]{20,}/, "GitHub token"],
  [/\bAKIA[0-9A-Z]{16}\b/, "AWS access key id"],
  [/\bxox[baprs]-[A-Za-z0-9-]{10,}/, "Slack token"],
  [/-----BEGIN (?:RSA |EC |OPENSSH |PGP )?PRIVATE KEY-----/, "private key"],
  [/\bey[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/, "JWT"],
];

type RecordValue = Record<string, unknown>;

function isRecord(value: unknown): value is RecordValue {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function record(value: unknown): RecordValue {
  return isRecord(value) ? value : {};
}

function issue(target: Issue[], pointer: string, message: string): void {
  target.push({ pointer, message });
}

function schemaIssues(errors: ErrorObject[] | null | undefined): Issue[] {
  return (errors ?? []).map((error) => ({ pointer: error.instancePath || "", message: error.message ?? "schema validation failed" }));
}

function* walkStrings(value: unknown, pointer = ""): Generator<[string, string]> {
  if (typeof value === "string") yield [pointer, value];
  else if (Array.isArray(value)) {
    for (const [index, child] of value.entries()) yield* walkStrings(child, `${pointer}/${index}`);
  } else if (isRecord(value)) {
    for (const [key, child] of Object.entries(value)) yield* walkStrings(child, `${pointer}/${key.replaceAll("~", "~0").replaceAll("/", "~1")}`);
  }
}

export function escapesWorkspace(path: string): boolean {
  if (path.startsWith("/") || /^[A-Za-z]:[\\/]/.test(path) || path.startsWith("\\\\")) return true;
  let depth = 0;
  for (const part of path.split(/[\\/]+/)) {
    if (!part || part === ".") continue;
    if (part === "..") {
      depth -= 1;
      if (depth < 0) return true;
    } else depth += 1;
  }
  return false;
}

function checkVersion(document: JsonObject, errors: Issue[]): boolean {
  if (typeof document.oap !== "string" || !/^\d+\.\d+$/.test(document.oap)) {
    issue(errors, "/oap", "missing or malformed spec version string");
    return false;
  }
  const [major, minor] = document.oap.split(".").map(Number);
  if (major !== 1 || (minor ?? 0) > 0) {
    issue(errors, "/oap", `unsupported OAP version ${document.oap}; unsupported versions fail closed`);
    return false;
  }
  return true;
}

function checkSecrets(document: JsonObject, errors: Issue[]): void {
  for (const [pointer, text] of walkStrings(document)) {
    for (const [pattern, label] of SECRET_PATTERNS) {
      if (pattern.test(text)) {
        issue(errors, pointer, `looks like a literal ${label}; use a \${VARIABLE} reference`);
        break;
      }
    }
  }
}

function checkProfile(document: JsonObject, errors: Issue[], warnings: Issue[], filename?: string): void {
  const spec = record(document.spec);
  const tools = record(spec.tools);
  const servers = Array.isArray(tools.mcp_servers) ? tools.mcp_servers : [];
  servers.forEach((server, index) => {
    if (!isRecord(server)) return;
    for (const [key, value] of Object.entries(record(server.env))) {
      if (typeof value !== "string" || !ENV_REF.test(value)) issue(errors, `/spec/tools/mcp_servers/${index}/env/${key}`, "must be a same-name ${VARIABLE} reference, not a literal");
    }
    for (const [key, value] of Object.entries(record(server.headers))) {
      if (typeof value !== "string" || !HEADER_REF.test(value)) issue(errors, `/spec/tools/mcp_servers/${index}/headers/${key}`, "must be '${VARIABLE}' or 'Bearer ${VARIABLE}'");
    }
  });

  const context = record(spec.context);
  const paths: Array<[string, string]> = [];
  (Array.isArray(context.files) ? context.files : []).forEach((entry, index) => {
    if (isRecord(entry) && typeof entry.path === "string") paths.push([`/spec/context/files/${index}/path`, entry.path]);
  });
  if (typeof context.working_directory === "string") paths.push(["/spec/context/working_directory", context.working_directory]);
  const filesystem = record(record(spec.permissions).filesystem);
  for (const key of ["read_roots", "write_roots", "deny_paths"]) {
    (Array.isArray(filesystem[key]) ? filesystem[key] : []).forEach((path, index) => {
      if (typeof path === "string") paths.push([`/spec/permissions/filesystem/${key}/${index}`, path]);
    });
  }
  for (const [pointer, path] of paths) if (escapesWorkspace(path)) issue(errors, pointer, `${JSON.stringify(path)} resolves outside the workspace`);

  const variables = new Set(Object.keys(record(context.variables)));
  const role = record(spec.role);
  const roleFields: Array<[string, string]> = [];
  if (typeof role.instructions === "string") roleFields.push(["/spec/role/instructions", role.instructions]);
  for (const key of ["objectives", "constraints"]) {
    (Array.isArray(role[key]) ? role[key] : []).forEach((value, index) => {
      if (typeof value === "string") roleFields.push([`/spec/role/${key}/${index}`, value]);
    });
  }
  for (const [pointer, text] of roleFields) {
    for (const match of text.matchAll(VAR_REF)) if (!variables.has(match[1] ?? "")) issue(errors, pointer, `references undefined variable ${JSON.stringify(match[1])}`);
  }
  for (const [pointer, text] of walkStrings(document.state ?? {}, "/state")) {
    if (VAR_REF.test(text)) issue(warnings, pointer, "contains a ${{ vars.* }} template; substitution never runs inside state");
    VAR_REF.lastIndex = 0;
  }

  const state = record(document.state);
  for (const key of ["facts", "preferences", "open_threads"]) {
    const seen = new Set<unknown>();
    (Array.isArray(state[key]) ? state[key] : []).forEach((entry, index) => {
      if (!isRecord(entry)) return;
      if (seen.has(entry.id)) issue(errors, `/state/${key}/${index}/id`, `duplicate id ${JSON.stringify(entry.id)}`);
      seen.add(entry.id);
    });
  }

  const metadata = record(document.metadata);
  if (metadata.trust !== undefined) issue(warnings, "/metadata/trust", "trust in the file must be discarded and recomputed from the discovery root");
  if (filename && typeof metadata.name === "string") {
    let stem = basename(filename);
    const suffix = PROFILE_SUFFIXES.find((candidate) => stem.endsWith(candidate));
    if (suffix) stem = stem.slice(0, -suffix.length);
    if (stem && stem !== metadata.name) issue(warnings, "/metadata/name", `${JSON.stringify(metadata.name)} does not match file name ${JSON.stringify(stem)}; metadata.name wins`);
  }
  const history: RecordValue[] = Array.isArray(document.history) ? (document.history as unknown[]).filter(isRecord) : [];
  const revisions = history.map((entry) => entry.revision).filter((value): value is number => Number.isInteger(value));
  if (revisions.some((value, index) => index > 0 && value < (revisions[index - 1] ?? value))) issue(errors, "/history", "entries must be ordered oldest first by revision");
  const current = metadata.revision;
  if (typeof current === "number" && revisions.length > 0 && (revisions.at(-1) ?? 0) > current) issue(errors, "/history", "newest history revision exceeds metadata.revision");

  const policy = typeof tools.policy === "string" ? tools.policy : "allowlist";
  if (policy === "inherit" && (tools.allow || tools.deny)) issue(warnings, "/spec/tools", "policy is 'inherit', so allow and deny are ignored");
  if (policy === "allowlist" && isRecord(document.spec) && tools.allow === undefined && Object.keys(tools).length > 0) issue(warnings, "/spec/tools", "allowlist has an empty allow list, so the agent gets no tools");
}

function checkDelta(document: JsonObject, errors: Issue[], warnings: Issue[]): void {
  (Array.isArray(document.operations) ? document.operations : []).forEach((operation, index) => {
    if (!isRecord(operation)) return;
    if (typeof operation.path !== "string" || !(operation.path === "/state" || operation.path.startsWith("/state/"))) issue(errors, `/operations/${index}/path`, "operation is outside /state; contract changes belong in proposals");
  });
  (Array.isArray(document.proposals) ? document.proposals : []).forEach((proposal, index) => {
    if (!isRecord(proposal) || typeof proposal.path !== "string") return;
    const path = proposal.path;
    if (["/spec/tools", "/spec/permissions", "/spec/memory", "/spec/runtime/subagents"].some((prefix) => path.startsWith(prefix)) && proposal.risk !== "high") {
      issue(warnings, `/proposals/${index}`, `${path} must be treated as high risk regardless of its declared risk`);
    }
  });
}

export interface ValidateOptions {
  filename?: string;
}

export function validateOap(document: JsonObject, options: ValidateOptions = {}): ValidationReport {
  const errors: Issue[] = [];
  const warnings: Issue[] = [];
  const kind = typeof document.kind === "string" ? document.kind : "unknown";
  if (!checkVersion(document, errors)) return { kind, document, errors, warnings, ok: false };
  if (kind === "AgentProfile") {
    if (!validateProfileSchema(document)) errors.push(...schemaIssues(validateProfileSchema.errors));
  } else if (kind === "AgentStateDelta") {
    if (!validateDeltaSchema(document)) errors.push(...schemaIssues(validateDeltaSchema.errors));
  } else {
    issue(errors, "/kind", `${JSON.stringify(document.kind)} is not a known 1.x kind`);
    return { kind, document, errors, warnings, ok: false };
  }
  checkSecrets(document, errors);
  if (kind === "AgentProfile") checkProfile(document, errors, warnings, options.filename);
  else checkDelta(document, errors, warnings);
  const report: ValidationReport = { kind, document, errors, warnings, ok: errors.length === 0 };
  if (kind === "AgentProfile") report.digests = profileDigests(document as AgentProfile);
  return report;
}

export { deltaSchema, profileSchema };

File: rust/src/validate.rs
use std::{
    collections::BTreeSet,
    path::{Component, Path, PathBuf},
};

use regex::Regex;
use serde_json::Value;

use crate::{Digests, Document, Issue, ValidationReport, load, object, profile_digests, strings};

const PROFILE_SCHEMA: &str = include_str!("../schema/agent-profile.schema.json");
const DELTA_SCHEMA: &str = include_str!("../schema/agent-state-delta.schema.json");

fn issue(target: &mut Vec<Issue>, pointer: impl Into<String>, message: impl Into<String>) {
    target.push(Issue {
        pointer: pointer.into(),
        message: message.into(),
    });
}

/// Runs schema, semantic, and security validation on a parsed OAP document.
pub fn validate(document: &Document, filename: Option<&Path>) -> ValidationReport {
    let mut errors = vec![];
    let mut warnings = vec![];
    let version = document
        .get("oap")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let pieces: Vec<&str> = version.split('.').collect();
    if pieces.len() != 2 || pieces.iter().any(|part| part.parse::<u64>().is_err()) {
        issue(
            &mut errors,
            "/oap",
            "missing or malformed spec version string",
        );
    } else if pieces != ["1", "0"] {
        issue(
            &mut errors,
            "/oap",
            format!("unsupported OAP version {version}; unsupported versions fail closed"),
        );
    }
    let kind = document
        .get("kind")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if !matches!(kind, "AgentProfile" | "AgentStateDelta") {
        issue(
            &mut errors,
            "/kind",
            format!("{kind:?} is not a known 1.x kind"),
        );
    }
    let schema_text = if kind == "AgentStateDelta" {
        DELTA_SCHEMA
    } else {
        PROFILE_SCHEMA
    };
    if let Ok(schema) = serde_json::from_str(schema_text) {
        if let Ok(validator) = jsonschema::validator_for(&schema) {
            let instance = Value::Object(document.clone());
            for error in validator.iter_errors(&instance) {
                issue(
                    &mut errors,
                    error.instance_path().to_string(),
                    error.to_string(),
                );
            }
        }
    }
    check_literal_secrets(Value::Object(document.clone()), "", &mut errors);
    if kind == "AgentProfile" {
        check_profile(document, filename, &mut errors, &mut warnings);
    } else if kind == "AgentStateDelta" {
        check_delta(document, &mut errors, &mut warnings);
    }
    let digests: Option<Digests> = if kind == "AgentProfile" && errors.is_empty() {
        profile_digests(document).ok()
    } else {
        None
    };
    ValidationReport {
        kind: kind.into(),
        document: Some(document.clone()),
        ok: errors.is_empty(),
        errors,
        warnings,
        digests,
    }
}

/// Loads and validates an OAP document, returning parse failures as issues.
pub fn validate_path(path: impl AsRef<Path>) -> ValidationReport {
    let path = path.as_ref();
    match load(path) {
        Ok(document) => validate(&document, Some(path)),
        Err(error) => ValidationReport {
            kind: String::new(),
            document: None,
            errors: vec![Issue {
                pointer: String::new(),
                message: error.to_string(),
            }],
            warnings: vec![],
            digests: None,
            ok: false,
        },
    }
}

fn walk_strings(value: &Value, pointer: &str, visit: &mut impl FnMut(&str, &str)) {
    match value {
        Value::Object(map) => {
            for (key, child) in map {
                walk_strings(
                    child,
                    &format!("{pointer}/{}", key.replace('~', "~0").replace('/', "~1")),
                    visit,
                );
            }
        }
        Value::Array(items) => {
            for (index, child) in items.iter().enumerate() {
                walk_strings(child, &format!("{pointer}/{index}"), visit);
            }
        }
        Value::String(text) => visit(text, pointer),
        _ => {}
    }
}

fn check_literal_secrets(value: Value, pointer: &str, errors: &mut Vec<Issue>) {
    let patterns = [
        (r"AKIA[0-9A-Z]{16}", "AWS access key"),
        (r"gh[pousr]_[A-Za-z0-9_]{20,}", "GitHub token"),
        (r"sk-[A-Za-z0-9]{20,}", "API key"),
        (r"-----BEGIN [A-Z ]*PRIVATE KEY-----", "private key"),
    ];
    let compiled: Vec<_> = patterns
        .into_iter()
        .map(|(pattern, label)| (Regex::new(pattern).unwrap(), label))
        .collect();
    walk_strings(&value, pointer, &mut |text, location| {
        for (pattern, label) in &compiled {
            if pattern.is_match(text) {
                issue(
                    errors,
                    location,
                    format!("looks like a literal {label}; use a ${{VARIABLE}} reference"),
                );
            }
        }
    });
}

/// Returns whether a relative path lexically escapes its workspace root.
pub fn escapes_workspace(path: &str) -> bool {
    let candidate = PathBuf::from(path);
    if candidate.is_absolute() {
        return true;
    }
    let mut depth = 0_i64;
    for component in candidate.components() {
        match component {
            Component::ParentDir => {
                depth -= 1;
                if depth < 0 {
                    return true;
                }
            }
            Component::Normal(_) => depth += 1,
            _ => {}
        }
    }
    false
}

fn check_profile(
    document: &Document,
    filename: Option<&Path>,
    errors: &mut Vec<Issue>,
    warnings: &mut Vec<Issue>,
) {
    let spec = object(document.get("spec"));
    let context = object(spec.get("context"));
    let env = Regex::new(r"^\$\{[A-Z][A-Z0-9_]{0,63}\}$").unwrap();
    let header = Regex::new(r"^(Bearer )?\$\{[A-Z][A-Z0-9_]{0,63}\}$").unwrap();
    for (index, raw) in object(spec.get("tools"))
        .get("mcp_servers")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .enumerate()
    {
        let server = object(Some(raw));
        for (key, value) in object(server.get("env")) {
            let text = value.as_str().unwrap_or_default();
            if !env.is_match(text) || text != format!("${{{key}}}") {
                issue(
                    errors,
                    format!("/spec/tools/mcp_servers/{index}/env/{key}"),
                    "must be a same-name ${VARIABLE} reference, not a literal",
                );
            }
        }
        for (key, value) in object(server.get("headers")) {
            if !value.as_str().is_some_and(|text| header.is_match(text)) {
                issue(
                    errors,
                    format!("/spec/tools/mcp_servers/{index}/headers/{key}"),
                    "must be '${VARIABLE}' or 'Bearer ${VARIABLE}'",
                );
            }
        }
    }
    for (index, raw) in context
        .get("files")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .enumerate()
    {
        if let Some(path) = object(Some(raw)).get("path").and_then(Value::as_str) {
            if escapes_workspace(path) {
                issue(
                    errors,
                    format!("/spec/context/files/{index}/path"),
                    format!("{path:?} resolves outside the workspace"),
                );
            }
        }
    }
    if let Some(directory) = context.get("working_directory").and_then(Value::as_str) {
        if escapes_workspace(directory) {
            issue(
                errors,
                "/spec/context/working_directory",
                format!("{directory:?} resolves outside the workspace"),
            );
        }
    }
    let variables: BTreeSet<String> = object(context.get("variables")).keys().cloned().collect();
    let variable = Regex::new(r"\$\{\{\s*vars\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}").unwrap();
    walk_strings(
        &Value::Object(document.clone()),
        "",
        &mut |text, pointer| {
            for capture in variable.captures_iter(text) {
                if !variables.contains(&capture[1]) {
                    issue(
                        errors,
                        pointer,
                        format!("references undefined variable {:?}", &capture[1]),
                    );
                }
                if pointer.starts_with("/state") {
                    issue(
                        warnings,
                        pointer,
                        "contains a ${{ vars.* }} template; substitution never runs inside state",
                    );
                }
            }
        },
    );
    let state = object(document.get("state"));
    for collection in ["facts", "preferences", "open_threads", "glossary"] {
        let mut ids = BTreeSet::new();
        for (index, raw) in state
            .get(collection)
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
            .enumerate()
        {
            if let Some(id) = object(Some(raw)).get("id").and_then(Value::as_str) {
                if !ids.insert(id) {
                    issue(
                        errors,
                        format!("/state/{collection}/{index}/id"),
                        format!("duplicate id {id:?}"),
                    );
                }
            }
        }
    }
    let metadata = object(document.get("metadata"));
    if metadata.contains_key("trust") {
        issue(
            warnings,
            "/metadata/trust",
            "trust in the file must be discarded and recomputed from the discovery root",
        );
    }
    if let (Some(path), Some(name)) = (filename, metadata.get("name").and_then(Value::as_str)) {
        let base = path
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or_default()
            .split('.')
            .next()
            .unwrap_or_default();
        if base != name {
            issue(
                warnings,
                "/metadata/name",
                format!("{name:?} does not match file name {base:?}; metadata.name wins"),
            );
        }
    }
    let history = document
        .get("history")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let mut previous = 0;
    for entry in &history {
        let revision = object(Some(entry))
            .get("revision")
            .and_then(Value::as_u64)
            .unwrap_or(0);
        if revision < previous {
            issue(
                errors,
                "/history",
                "entries must be ordered oldest first by revision",
            );
            break;
        }
        previous = revision;
    }
    if previous
        > metadata
            .get("revision")
            .and_then(Value::as_u64)
            .unwrap_or(0)
    {
        issue(
            errors,
            "/history",
            "newest history revision exceeds metadata.revision",
        );
    }
    let tools = object(spec.get("tools"));
    let policy = tools
        .get("policy")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if policy == "inherit"
        && (!strings(tools.get("allow")).is_empty() || !strings(tools.get("deny")).is_empty())
    {
        issue(
            warnings,
            "/spec/tools",
            "policy is 'inherit', so allow and deny are ignored",
        );
    }
    if policy == "allowlist" && strings(tools.get("allow")).is_empty() {
        issue(
            warnings,
            "/spec/tools",
            "allowlist has an empty allow list, so the agent gets no tools",
        );
    }
}

fn check_delta(document: &Document, errors: &mut Vec<Issue>, warnings: &mut Vec<Issue>) {
    for (index, raw) in document
        .get("operations")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .enumerate()
    {
        let path = object(Some(raw))
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or_default();
        if path != "/state" && !path.starts_with("/state/") {
            issue(
                errors,
                format!("/operations/{index}/path"),
                "operation is outside /state; contract changes belong in proposals",
            );
        }
    }
    for (index, raw) in document
        .get("proposals")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .enumerate()
    {
        let proposal = object(Some(raw));
        let path = proposal
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or_default();
        if [
            "/spec/tools",
            "/spec/permissions",
            "/spec/memory",
            "/spec/runtime/subagents",
        ]
        .iter()
        .any(|prefix| path.starts_with(prefix))
            && proposal.get("risk").and_then(Value::as_str) != Some("high")
        {
            issue(
                warnings,
                format!("/proposals/{index}"),
                format!("{path} must be treated as high risk regardless of its declared risk"),
            );
        }
    }
}
